working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1404,30 +1404,66 @@ To orchestrate this, type inference on patterns proceeds in three phases:
+    object, and not necessarily to try to force a certain answer.*
+
 2.  **Calculate the static type of the matched value.** A pattern always occurs
     in the context of some matched value. For pattern variable declarations,
     this is the initializer. For switches and if-case statements, it's the value
     being matched.
 
-    Using the pattern's type schema as a context type, infer missing types on
-    the value expression. This is the existing type inference rules on
-    expressions. It yields a complete static type for the matched value.
+    Using the pattern's type schema as a context type (if not in a matching
+    context), infer missing types on the value expression. This is the existing
+    type inference rules on expressions. It yields a complete static type for
+    the matched value. This process may also insert implicit coercions and casts
+    from `dynamic` in the matched value expression.
+
+    *For example:*
+
+    ```dart
+    T id<T>(T t) => t;
+    dynamic d = 'str';
+    var (double n, int Function(int) f, String s) = (1, id, d);
+    ```
+
+    *This generates a type schema of `(double, int Function(int), String)` from
+    the pattern. That type schema is applied to the initializer, which inserts
+    coercions and casts to become:*
+
+    ```dart
+    var (double n, int Function(int) f, String s) = (1.0, id<int>, d as String);
+    ```
 
 3.  **Calculate the static type of the pattern.** Using that value type, recurse
     through the pattern again downwards to the leaf subpatterns filling in any
-    holes in the type schema. When that completes, we now have a full static
-    type for the pattern and all of its subpatterns.
+    holes in the type schema. This process may also insert implicit coercions
+    and casts from `dynamic` when values flow into a pattern during matching.
 
-The full process only comes into play for pattern variable declarations and
-pattern assignment. For switch cases and if-case statements, the pattern context
-type schema is not used and no downwards inference is performed from the pattern
-to the matched value expression.
+    *For example:*
+
+    ```dart
+    T id<T>(T t) => t;
+    (T Function<T>(T), dynamic) record = (t, 'str');
+    var (int Function(int) f, String s) = record;
+    ```
 
-*It would be hard to apply inference from cases in a switch to the value since
-there are multiple cases and it's not clear how to unify that. Even in case-if
-statements, it's not clear that downwards inference is desirable, since the
-intent of the pattern is to ask a question about the matched object, and not
-necessarily to try to force a certain answer.*
+    *Since the right-hand is not a record literal, we can't use the pattern's
+    context type schema to insert coercions when the record is being created.
+    However, `(T Function<T>(T), dynamic)` is assignable to `(int Function(int),
+    String)`, so the declaration is valid. Instead, coercions are inserted after
```

Comment:
It's not clear that the record types are assignable. This is a case where we need to destruct, coerce, and then recreate the record value, which is something I haven't been able to sell. (Probably a wise choice too.)

What is the case here is that `T Function<T>(T)` is assignable to `int Function(int)`, and `dynamic` to `String`, and we *don't* check the assignability of the record type itself here, only the of the actual values extracted by destructuring.

Reply:
> It's not clear that the record types are assignable.

I actually can't find where assignability is defined _at all_. The language spec still seems to have the old implicit downcast verbiage. I didn't find any mention of assignability in "resources/type-system".

@leafpetersen, is there somewhere else I should be looking?

> This is a case where we need to destruct, coerce, and then recreate the record value,

I don't think so. We have a record object already. The record pattern will match it as long as it has the right shape and doesn't care about the field types at all. Then the record pattern destructures each field and sends them to the corresponding field subpatterns. It's when sending the destructured field values into the subpatterns that we insert the coercions. So we should never need to allocate a new coerced record.

Reply:
Assignability is defined (at least) here: https://github.com/dart-lang/language/blob/master/accepted/2.12/nnbd/feature-specification.md#assignability

Which is not useful, since it only covers `dynamic`. 

A more recent reference is from ["constructor tearoffs"](https://github.com/dart-lang/language/blob/master/accepted/2.15/constructor-tearoffs/feature-specification.md#:~:text=both%20implicit%20and%20explicit%20instantiation).
(which, btw., was where we specified that you can do `.call` and instantiation in the same implicit operation.)

It's not specified in terms of "assignability" though.
It is probably just assumed to be part of type inference, which ... is still not formally specified?

So, specification is lacking.
What the *effect* has so far been is the following:
If `e` has static type `T` in context type `C` then
* If `T` has a `call` method (instance or extension) and `C` is a function type or `Function`, then treat `e` as if it was `e.call` with a static type of the function signature of the `call` method.
* If `T` is a generic function type and `C` is a non-generic function type, try to infer type arguments `typeArgs` such that the instantiation `T<typeArgs>` is assignable to `C` (in the specific way we usually do that). If so, treat `e` as if it was `e<typeArgs>` with static type `T<typeArgs>`.
* If `T` is dynamic and `C` is a non-top-type, treat `e` as if it was `e as C` with static type `C`.

The first two can both apply if the `call` method is generic.
We then say that `T` is assignable to `C` if the conversion above gives a program which is well typed.

For records, we have to decide whether the record type `(A, B)` is assignable to `(B, C)` by any means other than those above.
By pushing in the context type, we get conversion on the individual fields of a record literal.
For a non-literal, the above conversions do not apply (unless someone extends a record type with a `call` method).

Which means that if we don't destructure and rebuild, then `(T Function<T>(T), dynamic)` cannot be assignable to `(int Function(int), String)`.

Reply:
> For records, we have to decide whether the record type `(A, B)` is assignable to `(B, C)` by any means other than those above.
> ...
> Which means that if we don't destructure and rebuild, then `(T Function<T>(T), dynamic)` cannot be assignable to `(int Function(int), String)`.

I don't want to have to destructure *and rebuild*, but I think it's fine to apply the coercions at the point of destructuring. That suggests to me that we don't want to define assignability for records specially. A record type R is assignable to T only if R is a subtype of T, as usual. That way, this is forbidden:

```dart
(int, int) record = (1 as dynamic, 2 as dynamic);
```

Otherwise, we would have to destructure and then rebuild a new record. But I think allowing this is reasonable:

```dart
var (int x, int y) = (1 as dynamic, 2 as dynamic);
```

Because we don't need to actually build any new record here. I believe the current proposal draws this line. It forbids the former because there's no special assignability rules for records. It allows the latter because the required type of a record pattern always uses `Object?` for the field types. So in the second example, there is no compile time error because `(dynamic, dynamic)` is assignable to `(Object?, Object?)` and likewise `dynamic` is assignable to `Object?` for each field.